src/algorithm/center-of-mass.hxx

```diff
@@ -295,74 +296,119 @@ namespace pinocchio
     typedef DataTpl<Scalar,Options,JointCollectionTpl> Data;
-    
+
     typedef typename Model::JointIndex JointIndex;
-    
+
     typedef typename Data::SE3 SE3;
     typedef typename Data::Inertia Inertia;
-    
+
     data.com[0].setZero();
     data.mass[0] = Scalar(0);
-      
+
     for(JointIndex i=1; i<(JointIndex)(model.njoints); ++i)
     {
       const typename Inertia::Scalar & mass = model.inertias[i].mass();
       const typename SE3::Vector3 & lever = model.inertias[i].lever();
-      
+
       data.mass[i] = mass;
       data.com[i].noalias() = mass*data.oMi[i].act(lever);
     }
-   
+
     // Backward step
     typedef JacobianCenterOfMassBackwardStep<Scalar,Options,JointCollectionTpl> Pass2;
     for(JointIndex i= (JointIndex) (model.njoints-1); i>0; --i)
     {
       Pass2::run(model.joints[i],data.joints[i],
                  typename Pass2::ArgsType(model,data,computeSubtreeComs));
     }
-    
+
     data.com[0] /= data.mass[0];
     data.Jcom /=  data.mass[0];
-    
+
     return data.Jcom;
   }
 
+  template<typename Scalar, int Options, template<typename,int> class JointCollectionTpl, typename ConfigVectorType>
+  inline void
+  jacobianSubtreeCenterOfMass(const ModelTpl<Scalar,Options,JointCollectionTpl> & model,
+                              DataTpl<Scalar,Options,JointCollectionTpl> & data,
+                              const Eigen::MatrixBase<ConfigVectorType> & q,
+                              const JointIndex & parentJointIdIn,
+                              typename DataTpl<Scalar,Options,JointCollectionTpl>::Matrix3x & JacobianOut)
+  {
+    forwardKinematics(model, data, q);
+    jacobianSubtreeCenterOfMass(model, data, parentJointIdIn, JacobianOut);
+  }
+
+  template<typename Scalar, int Options, template<typename,int> class JointCollectionTpl>
+  inline void
+  jacobianSubtreeCenterOfMass(const ModelTpl<Scalar,Options,JointCollectionTpl> & model,
+                              DataTpl<Scalar,Options,JointCollectionTpl> & data,
+                              const JointIndex & parentJointIdIn,
+                              typename DataTpl<Scalar,Options,JointCollectionTpl>::Matrix3x & JacobianOut)
+  {
+    typedef typename DataTpl<Scalar,Options,JointCollectionTpl>::Matrix6x Matrix6x;
+    assert(model.check(data) && "data is not consistent with model.");
+    assert((parentJointIdIn < model.njoints) && "Invalid joint id.");
+
+    if (parentJointIdIn == 0)
+    {
+      // Joint 0 is universe: this means that the user is asking for the entier model CoM jacobian.
+      JacobianOut = jacobianCenterOfMass(model, data);
+    }
+    else
+    {
+      // Iterate over all bodies of subtree.
+      for (JointIndex id : model.subtrees[parentJointIdIn])
+      {
+        // Get joint jacobian in world.
+        Matrix6x jointJacobian = Matrix6x::Zero(6, model.nv);
+        getJointJacobian(model, data, id, ReferenceFrame::LOCAL, jointJacobian);
```

Comment:
This code is not efficient. You may proceed without requiring any call to `getJointJacobian`.
I think you might derive an efficient algo from the existing `jacobianCenterOfMass` with very few minor changes.

Reply:
Do you think that you will have time to do this is the coming days ? If not I may try to have a look at it tonight.

Reply:
Tomorrow if you don’t mind.